### ISSUE TITLE
FastAP mini-batch sampling scheme

### ISSUE THREAD

**Issue by fcakir** [KevinMusgrave/pytorch-metric-learning#105]
Hi, 
The FastAP method is used alongside a listwise mini-batch sampling scheme, [also introduced in the FastAP paper](https://github.com/kunhe/FastAP-metric-learning), which is shown to be crucial for the performances of listwise objectives such as FastAP. For reproducibility (and fair comparison purposes), is there a plan to incorporate this sampling scheme? 

Thank you.

**Comment by KevinMusgrave** [comment]
@fcakir Thanks for bringing this up. I think it's worth adding and benchmarking. Some initial thoughts about implementation:

1. It should be a pytorch sampler class, i.e. it gets passed into a dataloader.
2. To initialize it, it will require the dataset labels, and it will be assumed that the labels are a 2D array, where rows correspond to elements, and columns correspond to the hierarchical labels.
3. It should probably have an option for which hierarchy level to base its sampling off of, for datasets that have more than 2 levels of labels.
4. The name of the sampler could simply be HierarchicalSampler.

**Comment by elias-ramzi** [comment]
Hi Kevin,

Is it possible to make the HierarchicalSampler a batch sampler rather than a Sampler ?

Thanks

**Comment by KevinMusgrave** [comment]
I think that would be fine. (I'm assuming you're talking about a sampler that returns a batch of indices, as described in the documentation for the [pytorch dataloader](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader))
